designs/0005-static-matrices.md

```diff
@@ -0,0 +1,252 @@
+included if this is not true.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, an `N x M` matrix in Stan is represented as an Eigen matrix holding
+a pointer to an underlying array of autodiff objects, each of those holding
+a pointer to the underlying autodiff object implementation.
+
+The basic Eigen type used by Stan is `Eigen::Matrix<var, R, C>`. For brevity,
+this will be referred to as a `matvar` type (a matrix containing var variables).
+There are no major differences between a matrix of vars or a vector of vars or
+a row vector of vars.
+
+This proposal will enable an `N x M` Stan matrix to be defined in terms of one
+autodiff object with two separate `N x M` containers, one for its values and
+one for its adjoints.
+
+This basic example of this type is `var_value<Eigen::MatrixXd>`. For brevity,
+these will be referred to as `varmat` types (a var that contains a matrix). Again,
+there are no major differences when the matrix is replaced by a vector or a row vector.
+
+The motivation for introducing the `varmat` type is speed. These types speed up
+computations in a number of ways:
+
+1. The values in a `varmat` are stored separately from the `adjoints` and can
+be accessed independently. In a `matvar` the values and adjoints are interleaved,
+and each value/adjoint pair can be stored in a different place in memory.
+Reading a value or adjoint in a `matvar` requires dereferencing a pointer to
+where the value/adjoint pair is stored and then accessing them. Because of how
+caches and memory work on modern CPUs, even if only the value or adjoint are
+needed, both will be loaded.
+
+2. Reading all `N x M` values and adjoints of a `varmat` requires one pointer
+dereference and `2 x N x M` double reads. Reading the values and adjoints
+of a similar `matvar` requires `N x M` pointer dereferences and `2 x N x M` reads.
+
+3. Memory is linear in `varmat`, and in the best case interleaved in `matvar`
+(though because every element is a pointer, in worse case `matvar` memory accesses
+are entirely random). The linear memory in `varmat` means that they can use
+vectorized CPU instructions.
+
+4. `varmat` is a more suitable data structure for autodiff variables on a GPU
+than `matvar` because the memory difficulties with pointer-chasing and strided
+accesses are more acute on GPUs than CPUs. Linear memory also helps make the
+communication between the CPU and GPU efficient.
+
+5. Because values and adjoints of a `varmat` are each stored in one place, only
+two arena allocations are required to build an `N x M` matrix. `matvar` requires
+`N x M` allocations.
+
+See [here](https://github.com/stan-dev/design-docs/pull/21#issuecomment-625352581)
+for performance tests on a small subset of gradient evaluations using `varmat` and
+`matvar` types.
+
+`matvar` has two advantages that `varmat` does not duplicate. These are listed
+here to motivate the backwards compatibility of this proposal with `matvar`:
```

Comment:
I don't really like using the shorthand `varmat` and `matvar` here. I think we should just say "An Eigen matrix with a scalar var type" and "a var type that uses Eigen matrices as storage."

Reply:
Ooof, that was hard to edit in, so I kept the shorthand. I think this notation has been handy during development, so I like it for the design document.

`varmat` also includes such things as a view into another `varmat`, or an autodiff type on a GPU, etc. It's sort of imprecise, but it's handy that way (and `varmat`/`matvar` shorthands now come after a section with C++ examples and talking specifically about the details of the two implementations)

Reply:
We only use varmat / matvar because we type it a 100 times a day :-P but for defining a tech spec idt we should use shorthand verbage like that

Reply:
Was re-reading the last doc comments, a very early reference to something like varmat and matvar was Bob: https://github.com/stan-dev/design-docs/pull/21#discussion_r418700776 . So maybe it makes sense to people outside too? Anyway, this one I want to roll with until it's a problem.